src/Sql/dbo/Stored Procedures/Organization_EnableCollectionEnhancements.sql

```diff
@@ -21,15 +21,34 @@ BEGIN
           AND [AccessAll] = 1;
 
     -- Step 3: For all OrganizationUsers with Manager role or 'EditAssignedCollections' permission update their existing CollectionUser rows and insert new rows with [Manage] = 1
-    -- and finally update all OrganizationUsers with Manager role to User role
-        -- Create a temporary table to store the OrganizationUsers with Manager role or 'EditAssignedCollections' permission
+    -- and finally update all OrganizationUsers with Manager role or Custom users with only 'EditAssignedCollections' or 'DeleteAssignedCollections' to User role
+        -- Create a temporary table to store the OrganizationUsers with Manager role or with 'EditAssignedCollections' or 'DeleteAssignedCollections' permissions
         SELECT ou.[Id] AS [OrganizationUserId],
-            CASE WHEN ou.[Type] = 3 THEN 1 ELSE 0 END AS [IsManager]
+            CASE WHEN ou.[Type] = 3 THEN 1 ELSE 0 END AS [IsManager],
+            CASE WHEN ou.[Type] = 4 AND
+                         ou.[Permissions] IS NOT NULL AND
+                         ISJSON(ou.[Permissions]) > 0 AND
+                         JSON_VALUE(ou.[Permissions], '$.editAssignedCollections') = 'true'
+                    THEN 1 ELSE 0 END AS [EditAssignedCollections],
+            CASE WHEN ou.[Type] = 4 AND
+                         ou.[Permissions] IS NOT NULL AND
+                         ISJSON(ou.[Permissions]) > 0 AND
+                         (
+                            SELECT COUNT(*)
+                            FROM OPENJSON(ou.[Permissions])
+                            WHERE [key] NOT IN ('editAssignedCollections', 'deleteAssignedCollections') AND [value] = 'true'
+                         ) = 0
+                    THEN 1 ELSE 0 END AS [AllOtherPermissionsFalse]
         INTO #TempUserManagers
         FROM [dbo].[OrganizationUser] ou
         WHERE ou.[OrganizationId] = @OrganizationId
-            AND (ou.[Type] = 3 OR (ou.[Permissions] IS NOT NULL
-                AND ISJSON(ou.[Permissions]) > 0 AND JSON_VALUE(ou.[Permissions], '$.editAssignedCollections') = 'true'));
+              AND (ou.[Type] = 3 OR
+                   (ou.[Type] = 4 AND
+                    ou.[Permissions] IS NOT NULL AND
+                    ISJSON(ou.[Permissions]) > 0 AND
+                    (JSON_VALUE(ou.[Permissions], '$.editAssignedCollections') = 'true' OR
+                     JSON_VALUE(ou.[Permissions], '$.deleteAssignedCollections') = 'true')
+               ));
```

Comment:
This appears to parse the JSON at least 4 times:
* 3x calls to JSON_VALUE
* 1x call to OPENJSON

...plus probably some level of parsing involved in `ISJSON` in order to validate the format.

My guess is that SQL probably isn't smart enough to cache this (a quick Google [seems to confirm this](https://dba.stackexchange.com/questions/332558/performance-implications-of-openjson-in-view)). This may be why it had such an outsized impact on performance.

Is there a way to rewrite this to minimise JSON queries; can you use OPENJSON to parse it once and then query those results? e.g. parse it and then JOIN it to your `OrganizationUser` query, instead of parsing it in nested queries for each value?

Alternatively, you could make this change to the code instead - per your closed server PR - and avoid parsing JSON in SQL altogether. At this stage the opt-in is pure SQL migrations so I'm OK to keep the changes here, just mentioning it for completeness.

The performance is probably not noticeable for individual opt-in but seems like it will cause more work when running it in bulk.

Reply:
Took your suggestion and created new PR here https://github.com/bitwarden/server/pull/3770